File: packages/senri_ffi_js/src/memory.ts
/*
 * Copyright (c) 2026 TT23XR Studio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Pointer, PTR_BRAND } from './pointer';
import { FFIError } from './errors';
import { getGlobalAdapter } from './globals';
import { getResourceRegistry } from './resource-registry';

export function alloc(size: number): Pointer {
  if (typeof size !== 'number' || size <= 0) throw new FFIError('alloc requires a positive size');
  const adapter = getGlobalAdapter();
  const result = adapter.alloc(size);
  getResourceRegistry().registerMemory(result);
  return new Pointer(result);
}

export function free(ptr: Pointer | any): void {
  const adapter = getGlobalAdapter();
  const data = ptr instanceof Pointer ? ptr._data : ptr;
  getResourceRegistry().unregisterMemory(data);
  adapter.free(data);
}

export function addressOf(buffer: ArrayBuffer | ArrayBufferView): Pointer {
  if (!buffer || typeof buffer !== 'object') throw new FFIError('addressOf requires ArrayBuffer or TypedArray');
  const adapter = getGlobalAdapter();
  const addr = adapter.addressOf(buffer);
  return new Pointer({ __ptr: addr, __buf: buffer, __size: buffer.byteLength || (buffer as any).length || 0, [PTR_BRAND]: true });
}

export function errno(): number {
  try {
    return getGlobalAdapter().getErrno();
  } catch {
    return 0;
  }
}

export function strerror(code: number = 0): string {
  try {
    return getGlobalAdapter().getStrerror(code);
  } catch {
    return '';
  }
}

File: packages/senri_ffi_py/src/senri_ffi/memory.py
from __future__ import annotations

from typing import Any

from senri_ffi.errors import FFIError
from senri_ffi.globals import get_global_adapter
from senri_ffi.pointer import Pointer
from senri_ffi.registry import get_resource_registry


def alloc(size: int) -> Pointer:
    if not isinstance(size, int) or size <= 0:
        raise FFIError("alloc requires a positive size")
    adapter = get_global_adapter()
    result = adapter.alloc(size)
    get_resource_registry().register_memory(result)
    return Pointer(result)


def free(ptr: Pointer | Any) -> None:
    adapter = get_global_adapter()
    data = ptr._data if isinstance(ptr, Pointer) else ptr
    get_resource_registry().unregister_memory(data)
    adapter.free(data)


def address_of(buffer: Any) -> Pointer:
    if buffer is None or not isinstance(buffer, (bytes, bytearray, memoryview)):
        raise FFIError("address_of requires bytes, bytearray, or memoryview")
    adapter = get_global_adapter()
    addr = adapter.address_of(buffer)
    return Pointer({
        "__ptr": addr,
        "__buf": buffer,
        "__size": len(buffer),
        "__senri_ptr__": True,
    })


def errno() -> int:
    try:
        return get_global_adapter().get_errno()
    except Exception:
        return 0


def strerror(code: int = 0) -> str:
    try:
        return get_global_adapter().get_strerror(code)
    except Exception:
        return ""
